NIFI-4199: Consistent proxy support across components

ijokarumawak opened the pull request:
This PR adds ProxyConfigService into following processors. However, due to the restriction of underlying libraries, SOCKS and SOCKS+Auth are partially supported.

Category|Processors|HTTP|HTTP+Auth|SOCKS|SOCKS+Auth
---------|-----------|------|------------|-------|-------------|
FTP|FetchFTP GetFTP ListFTP PutFTP|OK|OK|OK|N/A
SFTP (jsch)|FetchSFTP GetSFTP ListSFTP PutSFTP|OK|OK|OK|OK
HTTP (HTTPClient)|GetHTTP PostHTTP|OK|OK|N/A|N/A
HTTP (OKHttp)|InvokeHTTP|OK|OK|OK|N/A
ES (OkHttp)|PutElasticsearchHttp QueryElasticsearchHttp FetchElasticsearchHttp FetchElasticsearchHttp PutElasticsearchHttpRecord|OK|OK|OK|N/A
AWS |PutS3Object ListS3 FetchS3Object DeleteS3Object PutKinesisFirehose PutKinesisStream PutLambda PutDynamoDB DeleteDynamoDB GetDynamoDB|OK|OK|N/A|N/A
Azure|FetchAzureBlobStorage ListAzureBlobStorage PutAzureBlobStorage DeleteAzureBlobStorage GetAzureQueueStorage PutAzureQueueStorage|OK|N/A|OK|N/A|

- Azure Storage client uses HTTPS, and since Azure SDK uses Java HttpURLConnection, we can not add authentication header from NiFi. The only way to set credential for Proxy auth is setting System properties. https://stackoverflow.com/questions/34877470/basic-proxy-authentication-for-https-urls-returns-http-1-0-407-proxy-authenticat/34980074
- Azure EventHub processors (ConsumeAzureEventHub, GetAzureEventHub, PutAzureEventHub) do not support proxy because it uses proton-j in Apache Qpid Proton. IOHandler.handleBound uses Java NIO SocketChannel, which doesn't support Proxy.





Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

MikeThomsen commented:
@ijokarumawak haven't had a chance to take a look at this, but have you tried it against Solr and Elastic yet? I think the latter's APIs do their own proxy management so that might need a little finessing here.

jugi92 commented:
It would be very nice if the initial proxy service also includes a SOCKS Proxy example. Other processors that implement the Proxy Service can then reuse the existing implementation even better. For example we would probably implement that change for the SFTP processors then.

ijokarumawak commented:
@ottobackwards I assume you were talking about this, #2016. That one adds user/password for proxy authentication at abstract AWS processor. This PR adds ProxyConfigurationService, which can be added on top of #2016 for AWS processors proxy configurations to be managed by the centralized Controller Service. Please look at the FTP and HTTP processors in this PR, AWS ones can adopt the CS same way.

ijokarumawak commented:
@jugi92 FTPTransfer supports SOCKS proxy. Specifically at these lines:
```
         if (proxyType == Proxy.Type.HTTP) {
-            client = new FTPHTTPClient(proxyHost, proxyPort, ctx.getProperty(HTTP_PROXY_USERNAME).getValue(), ctx.getProperty(HTTP_PROXY_PASSWORD).getValue());
+            client = new FTPHTTPClient(proxyHost, proxyPort, proxyConfig.getProxyUserName(), proxyConfig.getProxyUserPassword());
         } else {
             client = new FTPClient();
             if (proxyType == Proxy.Type.SOCKS) {
                 client.setSocketFactory(new SocksProxySocketFactory(new Proxy(proxyType, new InetSocketAddress(proxyHost, proxyPort))));
             }
         }
```
https://github.com/apache/nifi/pull/2704/files#diff-6e7e715d42f332cbe404edd9afbcaafaL533

For processors those don't support SOCKS proxy, following validation code should be added into their customValidate method, to confirm that ProxyConfigurationService is configured with the supported proxy type(s):
```
ProxyConfiguration.validateProxyType(validationContext, results, Proxy.Type.HTTP);
```

ProxyConfigurationService just holds the centralized proxy settings, each processor is responsible to use the settings with its own relying SDK/API way.

I checked #2018 but the PR doesn't look active. I will take a closer look on SFTP processor and #2018 to see if I can include SFTP ones into this PR, too.

ijokarumawak commented:
@MikeThomsen We can combine ProxyConfigurationService into ES or Solr, the CS just let users manage proxy settings in a centralized place. I will take a look on #2094 to see how I can help review that one. Thanks.

ijokarumawak commented:
Now this PR includes SFTP processors and SOCKS proxy support for SFTP as well.

ijokarumawak commented:
Elasticsearch processors are also included in this PR now.
I'm researching on AWS and Azure processors now, but those can be done separately.

jugi92 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/SFTPTransfer.java:
```diff
@@ -418,6 +424,23 @@ protected ChannelSftp getChannel(final FlowFile flowFile) throws IOException {
+                    break;
+                case SOCKS:
+                    final ProxySOCKS5 proxySOCKS5 = new ProxySOCKS5(proxyConfig.getProxyServerHost(), proxyConfig.getProxyServerPort());
+                    session.setProxy(proxySOCKS5);
```
can we add:
`
if (proxyConfig.hasCredential()) {
    socksProxy.setUserPasswd(proxyConfig.getProxyUserName(), proxyConfig.getProxyUserPassword());  
}
`

ijokarumawak replied:
@jugi92 Thanks! I actually didn't know SOCKS protocol supports user authentication.. It seems SFTP processors (thanks to the underlying jsch) support that. I've added your suggestion and confirmed with Dante SOCKS server with authentication.
https://gist.github.com/ijokarumawak/b3a31378bdc0a6c6b9922a138e9ec9c1

I will update this PR shortly.

ottobackwards commented:
@ijokarumawak I'm talking about passing around an HttpClientBuilder when not everyone uses that.

ijokarumawak commented:
Now this PR also includes AWS related processors. I've tested following processors can utilize HTTP forward proxies and support authentication:

- PutS3Object
- ListS3
- FetchS3Object
- DeleteS3Object
- PutKinesisFirehose
- PutKinesisStream
- PutLambda
- PutDynamoDB
- DeleteDynamoDB
- GetDynamoDB

ijokarumawak commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/s3/TestListS3.java:
```diff
@@ -311,5 +312,10 @@ public void testGetPropertyDescriptors() throws Exception {
         assertTrue(pd.contains(ListS3.USE_VERSIONS));
         assertTrue(pd.contains(ListS3.MIN_AGE));
+        assertTrue(pd.contains(ProxyConfigurationService.PROXY_CONFIGURATION_SERVICE));
+        assertTrue(pd.contains(ListS3.PROXY_HOST));
```
@jvwing Good catch, thanks. I removed the duplicated PROXY_HOST and PROXY_HOST_PORT. The missing one was LIST_TYPE. So, 17 + 3 = 20. 3 additions are PROXY_USER, PROXY_PASS and LIST_TYPE.

ijokarumawak commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/AbstractAWSProcessor.java:
```diff
@@ -92,6 +94,23 @@
             .build();
 
+    public static final PropertyDescriptor PROXY_USERNAME = new PropertyDescriptor.Builder()
+            .name("Proxy Username")
```
I will update it.

ijokarumawak commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/AbstractAWSProcessor.java:
```diff
@@ -92,6 +94,23 @@
+            .build();
+
+    public static final PropertyDescriptor PROXY_PASSWORD = new PropertyDescriptor.Builder()
+            .name("Proxy Password")
```
I will update it.

ijokarumawak commented:
Added 3 more commits.

1. Added proxy support to Azure processors.
2. Adding more explicit Proxy spec check and doc. Due to the restrictions of underlying libraries, Proxy support spec varies. Based on the investigation summarized in this PR's description, I've used 4 labels to represent spec HTTP, HTTP_AUTH, SOCKS and SOCKS_AUTH.
3. Incorporated review comments.

Example screenshots:
InvokeHTTP does not support SOCKS_AUTH, so if ProxyConfigurationService is configured with SOCKS and username/password, then it becomes invalid, but SOCKS without auth can be used:
![image](https://user-images.githubusercontent.com/1107620/40225356-c5e6a30c-5ac3-11e8-8bb4-9541144e6491.png)

PostHTTP does not support SOCKS at all:
![image](https://user-images.githubusercontent.com/1107620/40225479-1e24b64e-5ac4-11e8-880c-bb7fbe0642a7.png)

Not only validation, property description shows what proxy is supported:
![image](https://user-images.githubusercontent.com/1107620/40225514-37432408-5ac4-11e8-89df-654fcef8e8ef.png)

SFTP processors are the only ones supporting all Proxy specs:
![image](https://user-images.githubusercontent.com/1107620/40225546-4a64dbda-5ac4-11e8-8eae-b421f0fdb585.png)

ijokarumawak commented:
I really stop updating this PR. No more addition from my side. Let's wrap this up. Thanks for reviewing!

ottobackwards commented:
Should the tests for InvokeHTTP be updated to test with the changes?

MikeThomsen commented:
@trixpan @ijokarumawak There are two other tickets referenced, 4196 and 4175(?) in the commit list for this PR. Before I keep squashing, I want to confirm that you want me to keep going and put 3 "This closes #ABCD" statements in there to close this, 4196 and 4175.

ijokarumawak commented:
@MikeThomsen Thanks for merging this. Although my original intent was keeping commits made by @trixpan separated (not squashed) to retain his credits, it looks good to me because the original PR 4196 and 4175 are closed as I expected. @trixpan Thanks again for originating this improvements!

ottobackwards commented:
I found a bug in this in the aws implementation, I am not sure how you would see it in the other processors, I found it when bringing this code into my Gateway Api PR.

The issue is that customValidate validates that both host and port need to be set, but not that both user and password need to be set.

Since I test for this ( from the InvokeHttp testProxy ), I fail.

ottobackwards commented:
Once I prove out my fix and update my pr, I'll guess I'll do a PR against master with that fix?

ottobackwards commented:
https://issues.apache.org/jira/browse/NIFI-5220

ottobackwards commented:
https://github.com/apache/nifi/pull/2727
